Remove boxes from ast list elements

camsteffen opened the pull request:
Less indirection should be better perf.

camsteffen committed:
Remove boxes from ast Pat lists

camsteffen committed:
Remove boxes from ast TyPat lists

camsteffen commented:
Moved a `Box::new` to try and fix regressions. Profiling shows the regressions are in `parse_param_general::{closure#0}`.

ZuseZ4 commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued 4c3b5d3811bd68ed990231b0dcfaad8b5ab255af with parent 1ed3cd7030718935a5c5e5c8f6581f36d8be179f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1ed3cd7030718935a5c5e5c8f6581f36d8be179f&end=4c3b5d3811bd68ed990231b0dcfaad8b5ab255af).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

bors commented:
:umbrella: The latest upstream changes (presumably #146666) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ZuseZ4 commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (a956f2901a29aa752beccde4e9506ad62ed577b9): [comparison URL](https://perf.rust-lang.org/compare.html?start=3369e82c6bc03c5cdb66f730dba6f738b74c8e1d&end=a956f2901a29aa752beccde4e9506ad62ed577b9&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3369e82c6bc03c5cdb66f730dba6f738b74c8e1d&end=a956f2901a29aa752beccde4e9506ad62ed577b9&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.3%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.4%, -1.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-1.4%, -0.0%] | 2     |
| All ❌✅ (primary)                 | -0.2% | [-1.4%, 0.3%]  | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3369e82c6bc03c5cdb66f730dba6f738b74c8e1d&end=a956f2901a29aa752beccde4e9506ad62ed577b9&stat=max-rss)
<details>
<summary>Results (primary -3.3%, secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.6%, 0.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [2.7%, 2.7%]   | 1     |
| Improvements ✅ <br /> (primary)   | -5.2% | [-8.2%, -2.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.7%, -2.7%] | 1     |
| All ❌✅ (primary)                 | -3.3% | [-8.2%, 0.6%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3369e82c6bc03c5cdb66f730dba6f738b74c8e1d&end=a956f2901a29aa752beccde4e9506ad62ed577b9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=3369e82c6bc03c5cdb66f730dba6f738b74c8e1d&end=a956f2901a29aa752beccde4e9506ad62ed577b9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 472.027s -> 470.099s (-0.41%)
**Artifact size**: 387.76 MiB -> 387.77 MiB (0.00%)

ZuseZ4 commented:
@bors2 try @rust-timer queue

rust-timer commented:
Queued 8474967d35fe23226085abfddaa3fcb76cc35dda with parent 5c7ae0c7ed184c603e5224604a9f33ca0e8e0b36, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5c7ae0c7ed184c603e5224604a9f33ca0e8e0b36&end=8474967d35fe23226085abfddaa3fcb76cc35dda).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

camsteffen committed:
Break out some parsing cold functions

This recovers some instruction count regressions after changing most Pat
parsing functions return a non-boxed Pat. It seems to be beneficial to
break out a #[cold] parsing recovery function when the function includes
more parsing, presumably because this requires more stack space and/or
mem copies when LLVM optimizes the wrong path.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18247540924/job/51957542619) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51957542619) [(plain)](https://github.com/rust-lang/rust/commit/80c3bd746db19d4da61e2cceea1bed228ba43e6c/checks/51957542619/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt check
Diff in /checkout/compiler/rustc_parse/src/parser/diagnostics.rs:1853:
     }
 
     #[cold]
-    fn recover_from_bad_qpath<T: RecoverQPath>(
-        &mut self,
-        base: T,
-    ) -> PResult<'a, T> {
+    fn recover_from_bad_qpath<T: RecoverQPath>(&mut self, base: T) -> PResult<'a, T> {
         if let Some(ty) = base.to_ty() {
             return self.maybe_recover_from_bad_qpath_stage_2(ty.span, ty);
         }
Diff in /checkout/compiler/rustc_parse/src/parser/diagnostics.rs:2946:
     }
 
     #[cold]
-    fn recover_unexpected_comma(
-        &mut self,
-        lo: Span,
-        rt: CommaRecoveryMode,
-    ) -> PResult<'a, ()> {
+    fn recover_unexpected_comma(&mut self, lo: Span, rt: CommaRecoveryMode) -> PResult<'a, ()> {
         // An unexpected comma after a top-level pattern is a clue that the
         // user (perhaps more accustomed to some other language) forgot the
         // parentheses in what should have been a tuple pattern; return a
fmt: checked 6451 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:47
  local time: Sat Oct  4 17:48:28 UTC 2025
  network time: Sat, 04 Oct 2025 17:48:28 GMT
```

</details>

camsteffen commented:
Let's try again.

ZuseZ4 commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

camsteffen commented:
@ZuseZ4 can you give it another try?

ZuseZ4 commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued c10c5b1ea7de995e6ec34ef6dcda49324b4c2810 with parent fed46ffd5059e11669df1bd9406b02914c3fb73f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=fed46ffd5059e11669df1bd9406b02914c3fb73f&end=c10c5b1ea7de995e6ec34ef6dcda49324b4c2810).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

camsteffen commented:
This looks like a small win now. I think the one regression is noise.

@rustbot ready

bors commented:
:pushpin: Commit 0ea65e0bcd98403bddf49ed4641ec47c32624ad5 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 0ea65e0bcd98403bddf49ed4641ec47c32624ad5 -->
<!-- homu: {"type":"Approved","sha":"0ea65e0bcd98403bddf49ed4641ec47c32624ad5","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/18548445144/job/52879429450)
Approved by: cjgillot
Pushing f5242367f4acf0130c9b68e6956531b2cb73bd38 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/18548445144/job/52879429450"},"merge_sha":"f5242367f4acf0130c9b68e6956531b2cb73bd38"} -->
